### PR TITLE
Add Flush Timeout to s3 Connector

### PR DESCRIPTION
**Pull request by saegel**
This PR adds the `flush_timeout` configuration option to the `S3Output` connector.
When the specified `flush_timeout` interval expires, the s3 connector will write all backlog elements, regardless whether the message backlog is actually full or not.
This prevents the connector from holding onto logs forever, if there are no new logs to process and the configured backlog size is not reached.

Changelog:

* Implement `flush_timeout` configuration option for the `S3Output` connector
* Add `setup` method where `_write_backlog` is scheduled
* Make `_s3_resource` a cached property
* Extend `setup` method to check whether the specified `bucket` is accessible
* Create `_handle_s3_error` decorator to handle all boto3 and boto-core related exceptions.

### CODE REVIEW

File: logprep/connector/s3/output.py [hunk 1]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
         }
         self._add_to_backlog(error_document, self._config.error_prefix)
         self._write_to_s3_resource()
+
+    def _add_dates(self, prefix):
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:266]: please add type hints

File: logprep/connector/s3/output.py [hunk 2]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+        date_format_matches = self._replace_pattern.findall(prefix)
+        if date_format_matches:
+            now = TimeParser.now()
+            for date_format_match in date_format_matches:
+                formatted_date = now.strftime(date_format_match[2:-1])
+                prefix = re.sub(date_format_match, formatted_date, prefix)
+        return prefix
+
+    @Metric.measure_time()
+    def _write_to_s3_resource(self):
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:276]: please add type hints

File: logprep/connector/s3/output.py [hunk 3]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+        """Writes a document into s3 bucket using given prefix."""
+        if self._backlog_size >= self._config.message_backlog_size:
+            self._write_backlog()
+
+    def _add_to_backlog(self, document: dict, prefix: str):
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:281]: please add type hint for return value
```suggestion
    def _add_to_backlog(self, document: dict, prefix: str) -> None:
```

File: logprep/connector/s3/output.py [hunk 4]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+        Parameters
+        ----------
+        document : dict
+           Document to store in backlog.
+        """
+        prefix = self._add_dates(prefix)
+        prefix = f"{self._base_prefix}{prefix}"
+        self._message_backlog[prefix].append(document)
+
+    def _write_backlog(self):
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:293]:
```suggestion
    def _write_backlog(self) -> None:
```

File: logprep/connector/s3/output.py [hunk 5]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+        """Write to s3 if it is not already writing."""
+        if not self._message_backlog:
+            return
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:296]: this branch is not covered by tests.
if it is needed, please add a test for this

File: logprep/connector/s3/output.py [hunk 6]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+            self._write_to_s3(document_batch, identifier)
+        except EndpointConnectionError as error:
+            raise FatalOutputError(self, "Could not connect to the endpoint URL") from error
+        except ConnectionClosedError as error:
+            raise FatalOutputError(
+                self,
+                "Connection was closed before we received a valid response from endpoint URL",
+            ) from error
+        except (BotoCoreError, ClientError) as error:
+            raise FatalOutputError(self, str(error)) from error
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:320]: this try except block is the same than in the setup. Consider to extract it in a decorator to implement a more DRYish way

File: logprep/connector/s3/output.py [hunk 7]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+            self._write_document_batch(document_batch, f"{prefix_mb}/{time()}-{uuid4()}")
+        self._message_backlog.clear()
+
+        if not self._config.call_input_callback:
+            return
+
+        if self.input_connector and hasattr(self.input_connector, "batch_finished_callback"):
+            self.input_connector.batch_finished_callback()
+
+    def _write_document_batch(self, document_batch: dict, identifier: str):
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:309]:
```suggestion
    def _write_document_batch(self, document_batch: dict, identifier: str) -> None:
```

File: logprep/connector/s3/output.py [hunk 8]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+            raise FatalOutputError(self, "Could not connect to the endpoint URL") from error
+        except ConnectionClosedError as error:
+            raise FatalOutputError(
+                self,
+                "Connection was closed before we received a valid response from endpoint URL",
+            ) from error
+        except (BotoCoreError, ClientError) as error:
+            raise FatalOutputError(self, str(error)) from error
+
+    def _write_to_s3(self, document_batch: dict, identifier: str):
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:322]:
```suggestion
    def _write_to_s3(self, document_batch: dict, identifier: str) -> None:
```

File: logprep/connector/s3/output.py [hunk 9]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+        self._logger.debug(f'Writing "{identifier}" to s3 bucket "{self._config.bucket}"')
+        s3_obj = self._s3_resource.Object(self._config.bucket, identifier)
+        s3_obj.put(Body=self._encoder.encode(document_batch), ContentType="application/json")
+        self.metrics.number_of_successful_writes += len(document_batch)
+
+    @staticmethod
+    def _build_no_prefix_document(message_document: dict, reason: str):
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:329]:
```suggestion
    def _build_no_prefix_document(message_document: dict, reason: str) -> None:
```

File: logprep/connector/s3/output.py [hunk 10]
```diff
@@ -327,3 +262,77 @@ def store_failed(self, error_message: str, document_received: dict, document_pro
...
+        document = {
+            "reason": reason,
+            "@timestamp": TimeParser.now().isoformat(),
+        }
+        try:
+            document["message"] = json.dumps(message_document)
```
Comment (ekneg54) [inline logprep/connector/s3/output.py:335]: consider to make this method not static and use the msgspec endcoder `self.encoder` for this. it is faster an should be the one consistent way to encode and decode json in this project

File: tests/unit/connector/test_opensearch_output.py [hunk 11]
```diff
@@ -352,7 +352,7 @@ def test_setup_raises_fatal_output_error_if_opensearch_error_is_raised(self):
             self.object.setup()
 
     def test_setup_registers_flush_timout_tasks(self):
-        # this test fails if opensearch is running on localhost
+        self.object._config.hosts = ["opensearch:9092"]
```
Comment (ekneg54) [inline tests/unit/connector/test_opensearch_output.py:355]: thank you

**Review by ekneg54 — APPROVED** [review]
thank you for your contribution.
I found some minor issues. Please have a look.

Mostly missing type hints. Please consider to scan the touched files for missing type hints I did not catch

**Review by ekneg54 — CHANGES_REQUESTED** [review]
sorry... wrong button. Please consider to fix the issues mentioned above

File: logprep/connector/s3/output.py (hunk 1 above)
Comment (clumsy9) [reply logprep/connector/s3/output.py]: Type hints were added.

File: logprep/connector/s3/output.py (hunk 5 above)
Comment (clumsy9) [reply logprep/connector/s3/output.py:308]: I wrote a simple test that calls `_write_backlog` on an empty message backlog queue (which might be the case if the timeout expires and no data was added to the queue).

File: logprep/connector/s3/output.py (hunk 6 above)
Comment (clumsy9) [reply logprep/connector/s3/output.py]: I have added the decorator `_handle_s3_error` that should handle all boto(3) releated exceptions for methods of `S3Output`.

File: logprep/connector/s3/output.py (hunk 10 above)
Comment (clumsy9) [reply logprep/connector/s3/output.py]: `_build_no_prefix_document` has become a method and uses the `encoder` for encoding.

**Review by ekneg54 — APPROVED** [review]

**Pull request merged**